### PR TITLE
Make it build with ghc-9.6

### LINKED ISSUE #3439: Please turn on -Wmissing-import-lists
I am using versions from CHaP, using the build plan specified by `cardano-node` - produced as a result of running:

```
nix build github:input-output-hk/cardano-node/master#project.x86_64-linux.plan-nix.json -o plan.json
```

relevant versions are here, which are the ones that I am using:

```
"cardano-ledger-binary-1.1.0.1-4dfc7b45882c15932e1b21e8ca21a140c225a8b30407267b8e602e6bff3681ed",
"cardano-ledger-byron-1.0.0.0-54be2730b8b028afc7e153fae3fc3793f0e58e30cbecf9086dcab8e78441d279",
```

I get build errors like this:

```
cardano-ledger-byron> src/Cardano/Chain/Update/Proposal.hs:188:18: error:
cardano-ledger-byron>     Ambiguous occurrence ‘annotation’
cardano-ledger-byron>     It could refer to
cardano-ledger-byron>        either the field ‘annotation’ of record ‘Annotated’,
cardano-ledger-byron>               imported from ‘Cardano.Ledger.Binary’ at src/Cardano/Chain/Update/Proposal.hs:62:3-16
cardano-ledger-byron>               (and originally defined in ‘cardano-ledger-binary-1.1.0.1:Cardano.Ledger.Binary.Decoding.Annotated’)
cardano-ledger-byron>            or the field ‘annotation’ of record ‘AProposal’,
cardano-ledger-byron>               defined at src/Cardano/Chain/Update/Proposal.hs:95:5
cardano-ledger-byron>     |
cardano-ledger-byron> 188 |   recoverBytes = annotation
cardano-ledger-byron>     |                  ^^^^^^^^^^
```

This is confusing, but would be far easier to diagnose if we would just keep explicit import lists.

### PR DESCRIPTION
**Pull request by erikd**
This should not be merged until I can remove the `source-repository-package` stanzas that I added to the `cabal.project` file.

Also fixes #3439

### CODE REVIEW

**Comment by erikd** [comment]
The `fourmolu` error is due to some needed `CPP`:
```
import Type.Reflection (TypeRep, pattern App, pattern Con
#if __GLASGOW_HASKELL__ < 906
    -- Ghc-9.6 removed the Fun constructor.
    , pattern Fun
#endif
    )
```
because the `Fun` pattern synonym was removed in GHC 9.6.

**Comment by erikd** [comment]
All the SRPs have been removed so this is not longer a draft/WIP.

File: cabal.project [hunk 1]
```diff
@@ -86,11 +87,13 @@ benchmarks: true
...
--- It's okay for us to have this since it's only our benchmarks
--- that depend on `weigh`, not our released packages.
-source-repository-package
-  type: git
-  location: https://github.com/fpco/weigh.git
-  tag: bfcf4415144d7d2817dfcb91b6f9a6dfd7236de7
-  --sha256: 01fy4nbq6kaqi73ydn6w7rd1izkg5p217q5znyp2icybf41sl1b6
+constraints:
+    -- Earlier versions have a slightly different set of type constraints.
+    persistent >= 2.14
```
Comment (lehins) [inline cabal.project]: Can we put this either
* under the `if impl(ghc >= 9.6)` below?
* or directly in the `ledger-state` package?

`ledger-state` package is the only one using persist and it's not even released to CHaPs, so it is a low impact problem.

File: eras/byron/ledger/impl/src/Cardano/Chain/Update/Proposal.hs
Comment (lehins) [inline eras/byron/ledger/impl/src/Cardano/Chain/Update/Proposal.hs:1]: Could you please bump up the version of this ~`cardano-ledger-binary`~ `cardano-ledger-byron` package to `1.0.0.1` and a changelog entry saying GHC-9.6 compatibility.

This is a non-semantic change, thus a minor patch, but it is also a changelog worthy change.

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs [hunk 2]
```diff
@@ -179,11 +179,11 @@ makeTx ::
...
 makeTx txBody keyPairs msigs = ShelleyTx txBody txWits . maybeToStrictMaybe
   where
+    txWits :: ShelleyTxWits (ShelleyEra c)
     txWits =
-      mempty
-        { addrWits = mkWitnessesVKey (hashAnnotated txBody) keyPairs
-        , scriptWits = msigs
-        }
+      mkBasicTxWits
+        & addrTxWitsL .~ (mkWitnessesVKey (hashAnnotated txBody) keyPairs)
```
Comment (lehins) [inline eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs]: Redundant parenthesis
```suggestion
        & addrTxWitsL .~ mkWitnessesVKey (hashAnnotated txBody) keyPairs
```

**Review by lehins — COMMENTED** [review]
Just a couple of minor comments, otherwise looks great. @erikd thank you for doing this work.

**Review by disassembler — APPROVED** [review]

File: eras/byron/ledger/impl/src/Cardano/Chain/Update/Proposal.hs
Comment (lehins) [reply eras/byron/ledger/impl/src/Cardano/Chain/Update/Proposal.hs:1]: @erikd I am really sorry!!! I meant to say `cardano-ledger-byron`, not `cardano-ledger-binary`. My muscle memory messed with my brain.

Comment (erikd) [reply eras/byron/ledger/impl/src/Cardano/Chain/Update/Proposal.hs:1]: Done.

**Pull request merged**